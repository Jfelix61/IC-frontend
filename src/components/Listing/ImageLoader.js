import React, { useState } from "react";
import PropTypes from "prop-types";

//Wrapper for image; changes class of image to loaded after lazy loading
const ImageLoader = ({ src, alt, className }) => {
  const [classNames, setClassNames] = useState(
    `${className ? className : ""} img-loading`
  );
  const onLoad = () => {
    setClassNames(`${className ? className : ""} img-loaded`);
  };

  return <img src={src} alt={alt} onLoad={onLoad} className={classNames} />;
};

ImageLoader.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  className: PropTypes.string
};

export default ImageLoader;
