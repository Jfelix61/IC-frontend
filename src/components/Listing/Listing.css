.listing {
  display: flex;
  flex-direction: column;
  overflow: hidden;

  background-color: #fff;

  border-radius: 5px;
}

.image-container {
  position: relative;
  display: flex;
  flex-direction: column;
}

.image-container > img {
  background-size: cover;
  object-fit: cover;

  width: 40rem;
  height: 25rem;
}

.image-container__details {
  position: absolute;
  z-index: 1;
  bottom: 0;
  color: white;
  width: 100%;
  text-align: center;
  transition: 0.2s;

  cursor: pointer;

  background-color: var(--details-bg-color);
}

.image-container__details:hover {
  background-color: var(--details-hover-bg-color);
}
