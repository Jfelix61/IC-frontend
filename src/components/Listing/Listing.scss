$bp-large: 68.75em; // 1100px

.listing {
  display: flex;
  flex-direction: column;
  overflow: hidden;

  background-color: #fff;

  border-radius: 5px;

  width: 40rem;
}

.image-container {
  position: relative;
  display: flex;
  flex-direction: column;
  background-color: lightgray;

  img {
    background-size: cover;
    object-fit: cover;

    width: 40rem;
    height: 26rem;

    @media screen and (max-width: var($bp-large)) {
      width: 100%;
      height: 100%;
    }
  }

  &__lazy-load {
    width: 40rem;
    height: 26rem;
  }

  .img-loading {
    opacity: 0;
  }

  .img-loaded {
    animation: FadeIn cubic-bezier(0.23, 1, 0.32, 1) 1;
    animation-duration: 0.7s;
  }

  &__details {
    position: absolute;
    z-index: 1;
    bottom: 0;
    color: white;
    width: 100%;
    text-align: center;
    transition: 0.2s;

    cursor: pointer;

    background-color: var(--details-bg-color);

    &:hover {
      background-color: var(--details-hover-bg-color);
    }
  }
}

@keyframes FadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.bid-details {
  padding: 0.5rem 1rem;
  display: flex;
  flex-direction: column;

  &__button {
    margin: 2rem 0 0.6rem 0;
    padding: 1rem;

    background-color: var(--primary-color);
    color: #fff;

    border: none;
    border-radius: 5px;

    text-transform: uppercase;
    font-weight: bold;
    font-family: inherit;
    font-size: 1.5rem;

    cursor: pointer;

    outline: none;

    transition: 0.2s;

    &:hover {
      transform: translateY(-2px);
      box-shadow: 0 0.2rem 1rem rgba(0, 0, 0, 0.2);
    }

    &--done {
      background-color: gray;
      cursor: default;

      &:hover {
        transform: none;
        box-shadow: none;
      }
    }
  }
}

.section {
  display: flex;
  justify-content: center;
  align-items: center;

  padding: 0.5rem 0;

  border-bottom: solid var(--gray) 2px;
}

.status {
  & > *:not(:last-child) {
    margin-right: 3rem;
  }

  &__container {
    display: flex;
    flex-direction: column;
    align-items: center;

    width: 13rem;

    & h6 {
      text-transform: uppercase;
      font-weight: 300;
      font-size: 1.1rem;
      margin-bottom: 0.3rem;
    }
  }

  &__divider {
    height: 4rem;
  }

  &__bid {
    color: var(--primary-color);
  }
}

.car-details {
  padding: 1.5rem 0;

  h4 {
    text-transform: uppercase;
    color: var(--dark-gray);

    width: 32rem;

    font-size: 1.5rem;
    text-align: center;
    font-weight: bold;
  }
}

.car-condition {
  & > *:not(:last-child) {
    margin-right: 3rem;
  }

  & > h4 {
    width: 12rem;
    padding: 0 2rem;
    font-weight: 400;
    text-align: center;
  }
}
