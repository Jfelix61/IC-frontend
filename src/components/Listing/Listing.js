import React, { useState } from "react";
import "./Listing.scss";
import Divider from "../Divider/Divider";
import Countdown from "../Countdown/Countdown";
import ImageLoader from "./ImageLoader";
import LazyLoad from "react-lazy-load";
import PropTypes from "prop-types";
import classnames from "classnames";

const Listing = ({ details }) => {
  const [bids, setBids] = useState(details.bids);
  const [isDone, setIsDone] = useState(false);

  const highestBid = bids.length
    ? bids[bids.length >= 0 ? bids.length - 1 : 0].amount
    : 0; //If length is 0 then the bid amount will be 0

  const addBid = () => {
    if (
      /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
        navigator.userAgent
      )
    )
      setBids([
        ...bids,
        {
          amount: highestBid + 250,
          dealership: "Instacarro",
          createdAt: new Date().toISOString(),
          channel: "Mobile"
        }
      ]);
    else
      setBids([
        ...bids,
        {
          amount: highestBid + 250,
          dealership: "Instacarro",
          createdAt: new Date().toISOString(),
          channel: "Web"
        }
      ]);
  };

  return (
    <div className="listing">
      <div className="image-container">
        <LazyLoad className="image-container__lazy-load" offset={0}>
          <ImageLoader src={details.imageUrl} alt="Car" />
        </LazyLoad>
        <h4 className="image-container__details ">ver detalhes</h4>
      </div>
      <div className="bid-details">
        <div className="section status">
          <div className="status__container">
            <h6>Tempo Restante</h6>
            <Countdown
              done={setIsDone}
              initialTimerTime={details.remainingTime}
            />{" "}
          </div>
          <Divider className="status__divider" />
          <div className="status__container">
            <h6>Ultima Oferta</h6>
            <h1 className="status__bid" data-testid="currentBid">
              R$ {highestBid.toLocaleString("pt-BR")}
            </h1>
          </div>
        </div>
        <div className="section car-details">
          <h4>
            {details.make} {details.model} {details.version} {details.year}
          </h4>
        </div>
        <div className="section car-condition">
          <h4>{details.year}</h4>
          <Divider className="status__divider" />
          <h4>{details.km.toLocaleString("pt-BR")} KM</h4>
        </div>
        <button
          onClick={isDone ? null : addBid}
          className={classnames(
            "bid-details__button",
            isDone ? "bid-details__button--done" : ""
          )}
        >
          Fazer oferta
        </button>
      </div>
    </div>
  );
};

Listing.propTypes = {
  details: PropTypes.shape({
    id: PropTypes.string,
    make: PropTypes.string.isRequired,
    model: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    version: PropTypes.string.isRequired,
    km: PropTypes.number.isRequired,
    remainingTime: PropTypes.number.isRequired,
    imageUrl: PropTypes.string.isRequired,
    bids: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.number.isRequired,
        dealership: PropTypes.string.isRequired,
        createdAt: PropTypes.string.isRequired,
        channel: PropTypes.oneOf(["Web", "Mobile"]).isRequired
      })
    ).isRequired
  }).isRequired
};

export default Listing;
