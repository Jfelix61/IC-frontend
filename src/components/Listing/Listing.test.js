import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";

import Listing from "./Listing";

afterEach(cleanup);

const mockDetails = {
  id: "5b83f2b0ce8dea001ac412ed",
  make: "HYUNDAI",
  model: "TUCSON",
  year: 2015,
  version: "2.0 MPFI GLS 16V 143CV 2WD FLEX 4P AUTOMATICO",
  km: 46460,
  remainingTime: 1456713,
  imageUrl:
    "https://uploads.instacarro.com/JPEG_20180827_1117185b83f2b0ce8dea001ac412ed-1048370064.jpg",
  bids: [
    {
      amount: 40000,
      dealership: "Instacarro",
      createdAt: "2018-08-27T14:40:38.478Z",
      channel: "Mobile"
    },
    {
      amount: 40250,
      dealership: "Instacarro",
      createdAt: "2018-08-27T14:43:11.925Z",
      channel: "Web"
    }
  ]
};

describe("Listing", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(<Listing details={mockDetails} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays the highest bid", () => {
    const { getByTestId } = render(<Listing details={mockDetails} />);

    expect(getByTestId("currentBid").textContent).toBe("R$ 40,250");
  });

  it("increases bid by R$ 250", () => {
    const { getByText, getByTestId } = render(
      <Listing details={mockDetails} />
    );
    fireEvent.click(getByText("Fazer oferta"));
    expect(getByTestId("currentBid").textContent).toBe("R$ 40,500");
  });
});
