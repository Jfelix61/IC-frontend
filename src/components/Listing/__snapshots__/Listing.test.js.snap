// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="listing"
  >
    <div
      class="image-container"
    >
      <div
        class="LazyLoad"
        style="height: 250px; width: 400px;"
      />
      <h4
        class="image-container__details "
      >
        ver detalhes
      </h4>
    </div>
    <div
      class="bid-details"
    >
      <div
        class="section status"
      >
        <div
          class="status__container"
        >
          <h6>
            Tempo Restante
          </h6>
          <h1
            class="countdown "
            data-testid="timerValue"
          >
            00:24:16
          </h1>
           
        </div>
        <div
          class="divider status__divider"
        >
          ​
        </div>
        <div
          class="status__container"
        >
          <h6>
            Ultima Oferta
          </h6>
          <h1
            class="status__bid"
            data-testid="currentBid"
          >
            R$ 40,250
          </h1>
        </div>
      </div>
      <div
        class="section car-details"
      >
        <h4>
          HYUNDAI TUCSON 2.0 MPFI GLS 16V 143CV 2WD FLEX 4P AUTOMATICO 2015
        </h4>
      </div>
      <div
        class="section car-condition"
      >
        <h4>
          2015
        </h4>
        <div
          class="divider status__divider"
        >
          ​
        </div>
        <h4>
          46,460 KM
        </h4>
      </div>
      <button
        class="bid-details__button"
      >
        Fazer oferta
      </button>
    </div>
  </div>
</DocumentFragment>
`;
